File: CS/UnboundSource/ViewModels/MainViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Windows.Input;
using DevExpress.Data;
using DevExpress.Mvvm;
using DevExpress.Mvvm.Native;

namespace UnboundSource.ViewModels {
    public class MainViewModel : ViewModelBase {
        private readonly ObservableCollection<Item> _items;
        private readonly Random _random = new();
        private int _columnCount = 2;
        public DevExpress.Data.UnboundSource UnboundSource { get => GetProperty(() => UnboundSource); set => SetProperty(() => UnboundSource, value); }

        public MainViewModel() {
            _items = new ObservableCollection<Item>(Enumerable.Range(0, 10)
                                                              .Select(i => new Item { Id = i, Name = $"Item {i}" }));

            _items.ForEach(item => item["Value 1"] = Convert.ToDouble(item.Id));

            UnboundSource = new DevExpress.Data.UnboundSource();

            UnboundSource.ValueNeeded += (_, e) => {
                if (EqualityComparer<string>.Default.Equals(e.PropertyName, nameof(Item.Id)))
                    e.Value = _items[e.RowIndex].Id;
                else if (EqualityComparer<string>.Default.Equals(e.PropertyName, nameof(Item.Name)))
                    e.Value = _items[e.RowIndex].Name;
                else if (EqualityComparer<string>.Default.Equals(e.PropertyName, nameof(Item.CreatedAt)))
                    e.Value = _items[e.RowIndex].CreatedAt;
                else
                    e.Value = _items[e.RowIndex][e.PropertyName];
            };

            UnboundSource.ValuePushed += (_, e) => {
                if (EqualityComparer<string>.Default.Equals(e.PropertyName, nameof(Item.Id)))
                    _items[e.RowIndex].Id = (int)e.Value;
                else if (EqualityComparer<string>.Default.Equals(e.PropertyName, nameof(Item.Name)))
                    _items[e.RowIndex].Name = (string)e.Value;
                else if (EqualityComparer<string>.Default.Equals(e.PropertyName, nameof(Item.CreatedAt)))
                    _items[e.RowIndex].CreatedAt = (DateTime)e.Value;
                else _items[e.RowIndex][e.PropertyName] = e.Value;
            };

            UnboundSource.Properties.Add(new UnboundSourceProperty(nameof(Item.Id),
                typeof(Item).GetProperty(nameof(Item.Id))?.PropertyType));
            UnboundSource.Properties.Add(new UnboundSourceProperty(nameof(Item.Name),
                typeof(Item).GetProperty(nameof(Item.Name))?.PropertyType));
            UnboundSource.Properties.Add(new UnboundSourceProperty("Value 1", typeof(double)));

            UnboundSource.SetRowCount(_items.Count);

            AddColumnCommand = new DelegateCommand(AddColumn);
        }

        private void AddColumn() {
            _items.ForEach(item => item[$"Value {_columnCount}"] = _random.NextDouble());
            UnboundSource?.Properties.Add(new UnboundSourceProperty($"Value {_columnCount}", typeof(double)));
            UnboundSource?.SetRowCount(_items.Count);

            _columnCount++;
        }

        public ICommand AddColumnCommand { get; }
    }

    public class Item : BindableBase {
        private readonly Dictionary<string, object> _customFieldValues = new();
        public DateTime CreatedAt { get => GetProperty(() => CreatedAt); set => SetProperty(() => CreatedAt, value); }
        public int Id { get => GetProperty(() => Id); set => SetProperty(() => Id, value); }
        public string Name { get => GetProperty(() => Name); set => SetProperty(() => Name, value); }

        public object this[string fieldName] {
            get {
                _customFieldValues.TryGetValue(fieldName, out var value);
                return value;
            }

            set {
                _customFieldValues[fieldName] = value;
                RaisePropertyChanged(fieldName);
            }
        }
    }
}

File: VB/UnboundSource/ViewModels/MainViewModel.vb
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Linq
Imports System.Runtime.CompilerServices
Imports DevExpress.Data
Imports DevExpress.Mvvm.CodeGenerators
Imports DevExpress.Mvvm.Native

Namespace UnboundSource.ViewModels

    <GenerateViewModel>
    Public Partial Class MainViewModel

        Private ReadOnly _items As ObservableCollection(Of Item)

        Private ReadOnly _random As Random = New(_, _)()

        Private _columnCount As Integer = 2

        <GenerateProperty>
        Private _unboundSource As DevExpress.Data.UnboundSource

        Public Sub New()
            _items = New ObservableCollection(Of Item)(Enumerable.Range(0, 10).[Select](Function(i) New Item With {.Id = i, .Name = $"Item {i}"}))
            _items.ForEach(Sub(item) item("Value 1") = Convert.ToDouble(item.Id))
            UnboundSource = New DevExpress.Data.UnboundSource()
            UnboundSource.ValueNeeded += Function(__, e)
                If EqualityComparer(Of String).Default.Equals(e.PropertyName, NameOf(Item.Id)) Then
                    e.Value = _items(e.RowIndex).Id
                ElseIf EqualityComparer(Of String).Default.Equals(e.PropertyName, NameOf(Item.Name)) Then
                    e.Value = _items(e.RowIndex).Name
                ElseIf EqualityComparer(Of String).Default.Equals(e.PropertyName, NameOf(Item.CreatedAt)) Then
                    e.Value = _items(e.RowIndex).CreatedAt
                Else
                    e.Value = _items(e.RowIndex)(e.PropertyName)
                End If
            End Function
            UnboundSource.ValuePushed += Function(__, e)
                If EqualityComparer(Of String).Default.Equals(e.PropertyName, NameOf(Item.Id)) Then
                    _items(e.RowIndex).Id = CInt(e.Value)
                ElseIf EqualityComparer(Of String).Default.Equals(e.PropertyName, NameOf(Item.Name)) Then
                    _items(e.RowIndex).Name = CStr(e.Value)
                ElseIf EqualityComparer(Of String).Default.Equals(e.PropertyName, NameOf(Item.CreatedAt)) Then
                    _items(e.RowIndex).CreatedAt = CDate(e.Value)
                Else
                    _items(e.RowIndex)(e.PropertyName) = e.Value
                End If
            End Function
            Properties.Add(New UnboundSourceProperty(NameOf(Item.Id), GetType(Item).GetProperty(NameOf(Item.Id))?.PropertyType))
            Properties.Add(New UnboundSourceProperty(NameOf(Item.Name), GetType(Item).GetProperty(NameOf(Item.Name))?.PropertyType))
            Properties.Add(New UnboundSourceProperty("Value 1", GetType(Double)))
            UnboundSource.SetRowCount(_items.Count)
        End Sub

        <GenerateCommand>
        Private Sub AddColumn()
            _items.ForEach(Sub(item) item($"Value {_columnCount}") = _random.NextDouble())
            UnboundSource?.Properties.Add(New UnboundSourceProperty($"Value {_columnCount}", GetType(Double)))
            UnboundSource?.SetRowCount(_items.Count)
            _columnCount += 1
        End Sub
    End Class

    <GenerateViewModel>
    Public Partial Class Item
        Implements INotifyPropertyChanged

        Private ReadOnly _customFieldValues As Dictionary(Of String, Object) = New(_, _)()

        <GenerateProperty>
        Private _createdAt As Date

        <GenerateProperty>
        Private _id As Integer

        <GenerateProperty>
        Private _name As String

        Default Public Property Item(ByVal fieldName As String) As Object
            Get
                Dim value As Object = Nothing
                _customFieldValues.TryGetValue(fieldName, value)
                Return value
            End Get

            Set(ByVal value As Object)
                _customFieldValues(fieldName) = value
                OnPropertyChanged(fieldName)
            End Set
        End Property

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Protected Overridable Sub OnPropertyChanged(<CallerMemberName> ByVal Optional propertyName As String = Nothing)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub
    End Class
End Namespace
